components/typeahead/typeahead.directive.ts

```diff
@@ -58,82 +58,96 @@ export class Typeahead implements OnInit {
               private loader:DynamicComponentLoader) {
   }
 
-  public get matches() {
-    return this._matches;
-  }
+  private asyncActions() {
+    let s = new Subject();
+    s.subscribe(() => {
+      this.typeahead
+        .subscribe(
+          (matches:string[]) => {
+            this._matches = matches.slice(0, this.typeaheadOptionsLimit);
+            this.finalizeAsyncCall();
+          },
+          (err:any) => {
+            console.error(err);
+          }
+        );
+    });
 
-  private debounce(func:Function, wait:number):Function {
-    let timeout:any;
-    let args:Array<any>;
-    let timestamp:number;
-    let waitOriginal:number = wait;
-
-    return function () {
-      // save details of latest call
-      args = [].slice.call(arguments, 0);
-      timestamp = Date.now();
-
-      // this trick is about implementing of 'typeaheadWaitMs'
-      // in this case we have adaptive 'wait' parameter
-      // we should use standard 'wait'('waitOriginal') in case of
-      // popup is opened, otherwise - 'typeaheadWaitMs' parameter
-      wait = this.container ? waitOriginal : this.typeaheadWaitMs;
-
-      // this is where the magic happens
-      let later = function () {
-
-        // how long ago was the last call
-        let last = Date.now() - timestamp;
-
-        // if the latest call was less that the wait period ago
-        // then we reset the timeout to wait for the difference
-        if (last < wait) {
-          timeout = setTimeout(later, wait - last);
-          // or if not we can null out the timer and run the latest
-        } else {
-          timeout = null;
-          func.apply(this, args);
+    this.keyUpEventEmitter
+      .debounceTime(100)
+      .subscribe(
```

Comment:
Wouldn't it be better to use flatMap() here? Like this:

```
this.keyUpEventEmitter
  .debounceTime(100)
  .flatMap(() => this.typeahead)
  .subscribe(
    (matches:string[]) => {
      this._matches = matches.slice(0, this.typeaheadOptionsLimit);
      this.finalizeAsyncCall();
    },
    (err:any) => {
      console.error(err);
    }
  );
```

Additionally we could add the `distinctUntilChanged()`-operator to prevent multiple async calls for the same input.